fem/libceed/ceed.cpp

```diff
@@ -253,19 +348,68 @@ static void InitCeedTensorBasisAndRestriction(const FiniteElementSpace &fes,
                              tp_el_dof.GetData(), restr);
 }
 
+static void InitCeedTensorBasisAndRestriction(const FiniteElementSpace &fes,
+                                              const IntegrationRule &ir,
+                                              Ceed ceed, CeedBasis *basis,
+                                              CeedElemRestriction *restr)
+{
+   InitCeedTensorBasis(fes, ir, ceed, basis);
+   InitCeedTensorRestriction(fes, ir, ceed, restr);
+}
+
 void InitCeedBasisAndRestriction(const FiniteElementSpace &fes,
                                  const IntegrationRule &irm,
                                  Ceed ceed, CeedBasis *basis,
                                  CeedElemRestriction *restr)
 {
-   if (UsesTensorBasis(fes))
+   // Check for FES -> basis, restriction in hash tables
+   const Mesh *mesh = fes.GetMesh();
+   const FiniteElement *fe = fes.GetFE(0);
+   const int P = fe->GetDof();
+   const int Q = irm.GetNPoints();
+   const int nelem = mesh->GetNE();
+   CeedBasisKey basis_key (const_cast<FiniteElementSpace *>(&fes),
+                           const_cast<IntegrationRule *>(&irm),
+                           P, Q);
+   auto basis_itr = internal::ceed_basis_map.find(basis_key);
+   CeedRestrKey restr_key (const_cast<FiniteElementSpace *>(&fes),
```

Comment:
Instead of casting you should just change the type of your keys to `const FiniteElementSpace*` and `const IntegrationRule*`

Reply:
Yeah that's better, and an obvious change I should have seen.